test/e2e/node/pre_stop.go

```diff
@@ -219,6 +219,67 @@ var _ = SIGDescribe("PreStop", func() {
 			return false, err
 		})
 	})
+
+	It("should block the Pod get terminated until preStop process finishes within termination grace period", func() {
+		gracefulTerminationPeriodSeconds := int64(60)
+		preStopProcessSec := 30
+		start := time.Now()
+		preStopFinish := start.Add(time.Second * time.Duration(preStopProcessSec))
+
+		By("creating the pod")
+		name := "pod-prestop-hook-" + string(uuid.NewUUID())
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{Name: name},
+			Spec: v1.PodSpec{
+				RestartPolicy: v1.RestartPolicyNever,
+				Containers: []v1.Container{
+					{
+						Name:    name,
+						Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+						Command: []string{"sh", "-c", "sleep 10"},
+						Lifecycle: &v1.Lifecycle{
+							PreStop: &v1.Handler{
+								Exec: &v1.ExecAction{
+									Command: []string{"sh", "-c", fmt.Sprintf("sleep %d", preStopProcessSec)},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		podClient.Create(pod)
+
+		By("waiting for pod running")
+		framework.ExpectNoError(f.WaitForPodRunning(pod.Name))
+
+		var err error
+		pod, err = podClient.Get(pod.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred(), "failed to get scheduled preStop pod")
+
+		By("deleting the pod gracefully")
+		err = podClient.Delete(pod.Name, metav1.NewDeleteOptions(gracefulTerminationPeriodSeconds))
+		Expect(err).NotTo(HaveOccurred(), "failed to delete pod")
+
+		By("verifying the pod running state while preStop completes during graceful termination period")
+		// On specified interval, get preStop-pod status and check container status
+		err = wait.Poll(time.Second*3, time.Second*time.Duration(gracefulTerminationPeriodSeconds), func() (bool, error) {
+			now := time.Now()
+			kubeletPod, err := f.PodClient().Get(pod.Name, metav1.GetOptions{})
+			Expect(err).ShouldNot(HaveOccurred(), "Pod MUST NOT be terminated without finishing preStop process wihtin graceful termination period")
+
+			By("Pod terminating gracefully")
+			Expect(kubeletPod.ObjectMeta.DeletionTimestamp).ToNot(BeNil())
+
+			framework.Logf("Expect Pod Phase : %v\nContainerCreatedAt: %v\npreStopFinishBy: %v\nNow: %v\nElapsed: %v", kubeletPod.Status.Phase, start, preStopFinish, now, time.Since(start))
+			if time.Since(preStopFinish) < time.Since(now) {
```

Comment:
I am confused. What is the reason that time.Since results are compared?

Reply:
To make sure, Pods are not in Succeeded phase till `preStopFinish` time is crossed.
Currently, `preStopFinish` time is calculated since the container is created. (As there is not much considerable time diff between creation and deletion API call here)

`time.Since()` is used to compare elapsed time between `t(preStopFinish)` and `t(now/podStatusPolling)`. 
Ref logs: https://github.com/kubernetes/kubernetes/issues/55807#issuecomment-461083118 

Is there any better way to compare time?

Reply:
Thanks. I understand. looks good.

Reply:
Thanks for reviewing @k-toyoda-pi.